sdk/personalizer/Azure.AI.Personalizer/src/Models/RankProcessor.cs

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using Newtonsoft.Json;
+using Rl.Net;
+
+namespace Azure.AI.Personalizer
+{
+    /// <summary> The Rank Processor. </summary>
+    internal class RankProcessor
+    {
+        private readonly LiveModel _liveModel;
+        internal PolicyRestClient RestClient { get; }
+
+        /// <summary> Initializes a new instance of RankProcessor. </summary>
+        public RankProcessor(LiveModel liveModel)
+        {
+            this._liveModel = liveModel;
+        }
+
+        /// <summary> Submit a Personalizer rank request. Receives a context and a list of actions. Returns which of the provided actions should be used by your application, in rewardActionId. </summary>
+        /// <param name="options"> A Personalizer Rank request. </param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        public Response<PersonalizerRankResult> Rank(PersonalizerRankOptions options, CancellationToken cancellationToken = default)
+        {
+            if (String.IsNullOrEmpty(options.EventId))
+            {
+                options.EventId = Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);
+            }
+            
+            HashSet<string> excludedSet = new HashSet<string>(options.ExcludedActions);
+
+            // Store the original action list
+            List<PersonalizerRankableAction> originalActions = new List<PersonalizerRankableAction>();
+            List<PersonalizerRankableAction> rankableActions = new List<PersonalizerRankableAction>();
+            List<PersonalizerRankableAction> excludedActions = new List<PersonalizerRankableAction>();
+            int idx = 0;
+            foreach (var action in options.Actions)
+            {
+                action.Index = idx;
+                originalActions.Add(action);
+                if (excludedSet.Contains(action.Id))
+                {
+                    excludedActions.Add(action);
+                }
+                else
+                {
+                    rankableActions.Add(action);
+                }
+                ++idx;
+            }
```

Comment:
This logic should be unnecessary. The `Rank` requests supports a `Lock` action parameter to handle the excluded case.